src/linear_algebra/finite_dimensional.lean

```diff
@@ -196,6 +196,162 @@ begin
+      { fsplit; simp; assumption, },
+      { simpa, } },
+    { intros a h₁, dsimp, rw [dif_pos h₁], intro h₂,
+      split_ifs with h₃,
+      { refl, },
+      { rw [if_neg h₃, zero_smul, eq_self_iff_true, not_true] at h₂, contradiction, }, }, },
+  { refine ⟨z, z.2, _⟩, dsimp only [f], erw [dif_pos z.2, if_pos]; rwa [subtype.coe_eta] },
+end
+
+/--
+If a finset has cardinality larger than `findim + 1`,
+then there is a nontrivial linear relation amongst its elements,
+such that the coefficients of the relation sum to zero.
+-/
+lemma exists_nontrivial_relation_sum_zero_of_dim_succ_lt_card
+  [finite_dimensional K V] {t : finset V} (h : findim K V + 1 < t.card) :
+  ∃ f : V → K, ∑ e in t, f e • e = 0 ∧ ∑ e in t, f e = 0 ∧ ∃ x ∈ t, f x ≠ 0 :=
+begin
+  -- pick an element x₀ ∈ t,
+  have card_pos : 0 < t.card := lt_trans (nat.succ_pos _) h,
+  obtain ⟨x₀, m⟩ := (finset.card_pos.1 card_pos).bex,
+  -- apply the previous lemma to the other xᵢ - x₀,
+  let shift : V ↪ V := ⟨λ x, x - x₀, add_left_injective (-x₀)⟩,
+  let t' := (t.erase x₀).map shift,
+  have h' : findim K V < t'.card,
+  { simp only [t', card_map, finset.card_erase_of_mem m],
+    exact nat.lt_pred_iff.mpr h, },
+  -- to obtain a function `g`
+  obtain ⟨g, gsum, x₁, x₁_mem, nz⟩ := exists_nontrivial_relation_of_dim_lt_card h',
+  -- and then obtain `f` by translating back by `x₀`,
+  -- and setting the value of `f` at `x₀` to ensure `∑ e in t, f e = 0`.
+  let f : V → K := λ z, if z = x₀ then - ∑ z in (t.erase x₀), g (z - x₀) else g (z - x₀),
+  refine ⟨f, _ ,_ ,_⟩,
+  { simp [f],
+    conv_lhs { apply_congr, skip, rw [ite_smul], },
+    rw [finset.sum_ite],
+    conv { congr, congr, apply_congr, simp [filter_eq', m], },
+    conv { congr, congr, skip, apply_congr, simp [filter_ne'], },
+    rw [sum_singleton, neg_smul, add_comm, ←sub_eq_add_neg, sum_smul, ←sum_sub_distrib],
+    simp [←smul_sub],
+    change (∑ (x : V) in t.erase x₀, (λ e, g e • e) (shift x)) = 0,
+    rw ←sum_map _ shift,
+    exact gsum, },
+  { rw [← insert_erase m, sum_insert (not_mem_erase x₀ t)],
+    dsimp [f],
+    rw [if_pos rfl],
+    conv_lhs { congr, skip, apply_congr, skip, rw if_neg (show x ≠ x₀, from (mem_erase.mp H).1), },
+    exact neg_add_self _, },
+  { refine ⟨x₁ + x₀, _, _⟩,
+    { rw finset.mem_map at x₁_mem,
+      rcases x₁_mem with ⟨x₁, x₁_mem, rfl⟩,
+      rw mem_erase at x₁_mem,
+      simp only [x₁_mem, sub_add_cancel, function.embedding.coe_fn_mk], },
+    { dsimp only [f],
+      split_ifs with hx₁ hx₀,
+      { simp only [ne.def, neg_eq_zero],
+        simp only [add_left_eq_self] at hx₁, subst hx₁,
+        simp only [exists_prop, mem_erase, function.embedding.coe_fn_mk,
+          finset.mem_map, ne.def, sub_eq_zero] at x₁_mem,
+        rcases x₁_mem with ⟨x₁_mem_w, ⟨x₁_mem_h_left_left, x₁_mem_h_left_right⟩, x₁_mem_h_right⟩,
+        contradiction, },
+      { simpa using nz, } } },
+end
+
+lemma exists_pos_of_sum_zero_of_exists_nonzero {K : Type*} [linear_ordered_field K]
```

Comment:
Should this stay in this file?

Reply:
I've moved it. It's now stated for linearly ordered comm_rings. But it doesn't even need the ring structure.

Reply:
I've moved it to the `decidable_linear_ordered_cancel_comm_monoid` section of `big_operators`, so it no longer needs a ring structure.